design/kubernetes-friendly-oam.md

```diff
@@ -0,0 +1,482 @@
+resources would not be required by the OAM specification. Runtimes could choose
+to iterate over the referenced components, render them in memory, and act upon
+the rendered components directly.
+
+### Traits
+
+Like workloads, each trait is of a distinct kind. A TraitDefinition defines the
+schema of each kind of trait, and limits what workload kinds it applies to:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: TraitDefinition
+metadata:
+  name: manualscalertrait.core.oam.dev
+spec:
+  appliesToWorkloads:
+    - server.core.oam.dev/v1alpha1
+    - worker.core.oam.dev/v1alpha1
+    - task.core.oam.dev/v1alpha1
+  definitionRef:
+    name: manualscalertrait.core.oam.dev
+```
+
+The above TraitDefinition references a Kubernetes CustomResourceDefinition named
+manualscalertrait.core.oam.dev, designating custom resources defined by said CRD
+to be valid OAM traits. As with WorkloadDefinitions, OAM runtimes that are not
+powered by Kubernetes could opt to either implement the CRD concept, or
+introduce a new field to the TraitDefinition spec that defines a kind of trait
+and its schema. Application operators can discover available OAM trait kinds by
+running `kubectl get traitdefinitions`. Traits would most frequently be defined
+by OAM runtimes or infrastructure operators.
+
+This document proposes that trait configuration be applied by be embedding one
+or more of a particular kind of trait under the `traits` array of the component
+it should apply to. For example:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: ApplicationConfiguration
+metadata:
+  name: cool-example
+spec:
+  components:
+  - componentName: example-server
+    traits:
+    - trait:
+        apiVersion: core.oam.dev/v1alpha1
+        kind: ManualScalerTrait
+        spec:
+          replicaCount: 3
+```
+
+Note that the embedded trait resource supports Kubernetes-style object metadata,
+but it may be omitted when authoring an ApplicationConfiguration. Metadata such
+as names and labels may be filled automatically by the OAM runtime if a distinct
+instance of a trait configuration is created as part of processing an
+ApplicationConfiguration. As with workloads, Kubernetes runtimes may treat
+traits as an [embedded object]. Each trait resource is embedded beneath the key
+`trait`. This allows the spec to add other keys alongside `trait` in future
+without breaking compatibility.
+
+### Scopes
+
+This document proposes that application scopes, while maintaining their current
+purpose of grouping application components, function identically to traits from
+an API perspective. Available kinds of scope are defined via a ScopeDefinition,
+which also specifies whether a component may exist in multiple scopes of a
+particular kind:
+
+```yaml
+apiVersion: core.oam.dev/v1alpha1
+kind: ScopeDefinition
+metadata:
+  name: networkscope.core.oam.dev
+spec:
+  allowComponentOverlap: true
+  crdName: networkscope.core.oam.dev
```

Comment:
we'd better align with trait:

```
apiVersion: core.oam.dev/v1alpha1
kind: ScopeDefinition
metadata:
  name: networkscope.core.oam.dev
spec:
  allowComponentOverlap: true
  definitionRef:
     indexName: networkscope.core.oam.dev
```

and for non-K8s:
apiVersion: core.oam.dev/v1alpha1
kind: ScopeDefinition
metadata:
  name: networkscope.core.oam.dev
spec:
  allowComponentOverlap: true
  definition: |
     {
      "$schema": "http://json-schema.org/draft-07/schema#",
      "type": "object",
      "required": [
        "replicaCount"
      ],
      "properties": {
        "networkId": {
          "type": "string",
          "description": "subnet id of this network scope"
        }
      }
    }
```

```

Reply:
+1

Reply:
Yeah, this should have been changed to `definitionRef` too - will fix.